lib/Zonemaster/CLI.pm

```diff
@@ -517,13 +518,24 @@ sub add_fake_delegation {
         if ($ip) {
             push @{ $data{ $self->to_idn( $name ) } }, $ip;
         }
+        elsif (Zonemaster::Zone->new( { name => $domain } )->is_in_zone( $name )) {
+            say STDERR "--ns in zone nameservers must be a name/ip pair.";
+            exit( 1 );
+        }
         else {
             my $n = $self->to_idn( $name );
             my @ips = Net::LDNS->new->name2addr($n);
-            push @{ $data{$n} }, $_ for @ips;
+            if ( @ips ) {
+                push @{ $data{$n} }, $_ for @ips;
+            }
+            else {
+                say STDERR "--ns ${name} nameserver has no IP address.";
+                exit( 1 );
+            }
```

Comment:
I agree with @matsduf. There are 2 types of checks: syntactic and semantic. The cli (and the backend) should ONLY do syntactic checks. They should ensure that the test being requested has the correct syntax. So with a name server, the cli should just check that a paramater has been supplied to the --ns option.

Then the cli should hand over the test to the engine, which is responsible for SEMANTIC checks, such as:
1. Is the given name server name a valid DNS label?
2. If it is in-zone, was an IP address also provided?

The point I am making is that the CLI should do as little as possible. It should not resolve names to addresses itself. The CLI is just a wrapper that interfaces to the engine. In all cases, the engine should do all the testing and report any errors.

Similarly, the backend's API is just an interface to the engine. It should only check the syntax of the JSON request, and then pass it to the engine. Any problems with the name server or resolving it to addresses, etc, should be checked for, and reported by, the engine.

The current model is in my opinion wrong. The cli (and the backend) are doing some of the work themselves, and hiding errors from the engine. But this is also inconsistent, and confusing. I should be able to submit a check via the CLI or the backend, and expect the exact same result from the engine. And the only way for this to happen is for both the CLI and backend to just be wrappers for the engine, and not do any of the DNS checks themselves.